source/opt/dead_branch_elim_pass.cpp

```diff
@@ -0,0 +1,335 @@
+    def_use_mgr_->KillInst(ip);
+  });
+}
+
+bool DeadBranchElimPass::GetConstConditionalSelectionBranch(ir::BasicBlock* bp,
+    ir::Instruction** branchInst, ir::Instruction** mergeInst,
+    uint32_t *condId, bool *condVal) {
+  auto ii = bp->end();
+  --ii;
+  *branchInst = &*ii;
+  if ((*branchInst)->opcode() != SpvOpBranchConditional)
+    return false;
+  if (ii == bp->begin())
+    return false;
+  --ii;
+  *mergeInst = &*ii;
+  if ((*mergeInst)->opcode() != SpvOpSelectionMerge)
+    return false;
+  bool condIsConst;
+  *condId = (*branchInst)->GetSingleWordInOperand(
+      kBranchCondConditionalIdInIdx);
+  (void) GetConstCondition(*condId, condVal, &condIsConst);
+  return condIsConst;
+}
+
+bool DeadBranchElimPass::HasNonPhiRef(uint32_t labelId) {
+  analysis::UseList* uses = def_use_mgr_->GetUses(labelId);
+  if (uses == nullptr)
+    return false;
+  for (auto u : *uses)
+    if (u.inst->opcode() != SpvOpPhi)
+      return true;
+  return false;
+}
+
+bool DeadBranchElimPass::EliminateDeadBranches(ir::Function* func) {
+  // Traverse blocks in structured order
+  std::list<ir::BasicBlock*> structuredOrder;
+  ComputeStructuredOrder(func, &structuredOrder);
+  std::unordered_set<ir::BasicBlock*> elimBlocks;
+  bool modified = false;
+  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
+    // Skip blocks that are already in the elimination set
+    if (elimBlocks.find(*bi) != elimBlocks.end())
+      continue;
+    // Skip blocks that don't have constant conditional branch preceded
+    // by OpSelectionMerge
+    ir::Instruction* br;
+    ir::Instruction* mergeInst;
+    uint32_t condId;
+    bool condVal;
+    if (!GetConstConditionalSelectionBranch(*bi, &br, &mergeInst, &condId,
+        &condVal))
+      continue;
+
+    // Replace conditional branch with unconditional branch
+    const uint32_t trueLabId =
+        br->GetSingleWordInOperand(kBranchCondTrueLabIdInIdx);
+    const uint32_t falseLabId =
+        br->GetSingleWordInOperand(kBranchCondFalseLabIdInIdx);
+    const uint32_t mergeLabId =
+        mergeInst->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+    const uint32_t liveLabId = condVal == true ? trueLabId : falseLabId;
+    const uint32_t deadLabId = condVal == true ? falseLabId : trueLabId;
+    AddBranch(liveLabId, *bi);
+    def_use_mgr_->KillInst(br);
+    def_use_mgr_->KillInst(mergeInst);
+
+    // Iterate to merge block adding dead blocks to elimination set
+    std::unordered_set<uint32_t> deadSuccIds;
+    deadSuccIds.insert(deadLabId);
+    auto dbi = bi;
+    ++dbi;
+    uint32_t dLabId = (*dbi)->id();
+    while (dLabId != mergeLabId) {
+      if (deadSuccIds.find(dLabId) != deadSuccIds.end()) {
+        // Add successor blocks to dead successor set
+        (*dbi)->ForEachSuccessorLabel([&deadSuccIds](uint32_t succ) {
+          deadSuccIds.insert(succ);
+        });
+        // If dead block is merge block, add its merge block to dead
```

Comment:
"If dead block is a *header* block, ...."

By the same logic, shouldn't we also add its continue target if it has one?  If the header block of a construct is dead then all its nested constructs would be dead too.

Optional to fix, since it could be cleaned up  with a separate "remove dead blocks" pass.